File: vectordb.py
import chromadb
import json
from utils import generate_id,get_schema,generate_embeddings,get_questions

class VectorDB:
    def __init__(self, path:str):
        self.db = chromadb.PersistentClient(path=path)


    def read_questions(self):
        try:
            questions= get_questions("postgres","adit1290","localhost", "argo_db")
            for question_data in questions:
                print(question_data)
                self.add_vector(data=question_data,collection_name="questions")
            return True
        except Exception as e:
            return {
                "error":str(e)
            },False

    def read_schema(self):
        try:
            schema= get_schema("postgres","adit1290","localhost", "argo_db")
            for column_data in schema:
                print(column_data)
                self.add_vector(data=column_data,collection_name="schema")
            return True
        except Exception as e:
            return {
                "error":str(e)
            },False

    def add_vector(self,data:str,collection_name:str)->str:
        try:
            collection = self.db.get_or_create_collection(name=collection_name)
            id = generate_id(data=data) + "-" + collection_name
            embeddings = generate_embeddings(data=data)
            collection.add(
                    documents=data,
                    embeddings=embeddings,
                    ids=id
                )
            return {
                "id":id,
                "data":data,
                "collection":collection_name,
                "embeddings":embeddings
            },True
        except Exception as e:
            return {
                "error":str(e)
            },False
        
    def query(self,data:str,collection_name:str,n_results:int)->list:
        try:
            embeddings = generate_embeddings(data=data)
            collection = self.db.get_collection(name=collection_name)
            return self.return_documents(
            collection.query(
                    query_embeddings=embeddings,
                    n_results=n_results
                )),True
        except Exception as e:
           return {
                "error":str(e)
            },False 

    def return_documents(self,results) -> list:
        if results is None:
            return []
        
        if "documents" in results:
            documents = results["documents"]

            if len(documents) == 1 and isinstance(documents[0], list):
                try:
                    documents = [json.loads(doc) for doc in documents[0]]
                except Exception as e:
                    return documents[0]
            return documents

    def clear_collection(self,collection_name:str):
        try:
            self.db.delete_collection(name=collection_name)
            return {
                "collection":collection_name
            },True
        except Exception as e:
            return {
                "error":str(e)
            },False 

File: javascript/src/vectordb.js
import { ChromaClient } from "chromadb";
import { get_questions,generateId,generateEmbeddings,get_scehama } from './utils.js';

const myNomicEmbeddingFunction = {
    generate: async function (texts) {
        return await generateEmbeddings(texts);
    }
};

class VectorDB 
{
    constructor()
    {
        this.db = new ChromaClient();
    }

    async read_questions()
    {
        const questions = await get_questions();
        for (let i = 0; i < questions.length; i++)
        {
            console.log(questions[i]);
            await this.add_vector(`Question : ${questions[i]['question']}, Query : ${questions[i]['query']}` ,"questions");
        }
    }

    async read_schema()
    {
        const schema = await get_scehama();
        for (let i = 0; i < schema.length; i++)
        {
            console.log(schema[i]);
            await this.add_vector(schema[i],"schema");
        }
    }

    async add_vector(data,collection_name)
    {
        const collection = await this.db.getOrCreateCollection({
            name: collection_name,
            embeddingFunction: myNomicEmbeddingFunction,
        });
        await collection.add({
            documents: [data],
            ids: [generateId(data)],
        });
    }

    async query(data, collection_name, n_results)
    {
        const embeddings = generateEmbeddings(data);
        const collection = await this.db.getCollection({ name: collection_name,embeddingFunction: myNomicEmbeddingFunction });
        const results = await collection.query({
                queryTexts: [data],
                nResults: n_results,
        });
        const documents = this.return_documents(results);
        return documents;
    }

    return_documents(results) {
    if (!results || !results.documents) {
        return [];
    }

    let documents = results.documents;
    if (documents.length === 1 && Array.isArray(documents[0])) {
        return documents[0];
    }
    return documents;
}

    async clear_collection(collection_name)
    {
        try {
            await this.db.deleteCollection({ name: collection_name });
            return true;
        } catch (e) {
            return { error: e.message, success: false };
        }
    }
}

export { VectorDB };